### PR TITLE
Fix/refactor

### PR DESCRIPTION
**Pull request by yunzc**
Added stats reporting and also some clean up and refactoring

### CODE REVIEW

File: RobustPGO/RobustSolver.cpp [hunk 1]
```diff
@@ -94,7 +102,7 @@ void RobustSolver::optimize() {
 }
 
 void RobustSolver::update(const gtsam::NonlinearFactorGraph& nfg,
-                       const gtsam::Values& values) {
+                          const gtsam::Values& values) {
```
Comment (marcusabate) [inline RobustPGO/RobustSolver.cpp]: Make this the same convention as the rest of the codebase - arguments on the next line of the declaration should be indented four spaces over.

```suggestion
    const gtsam::Values& values) {
```

File: RobustPGO/RobustSolver.cpp [hunk 2]
```diff
@@ -111,7 +119,7 @@ void RobustSolver::update(const gtsam::NonlinearFactorGraph& nfg,
 }
 
 void RobustSolver::forceUpdate(const gtsam::NonlinearFactorGraph& nfg,
-                       const gtsam::Values& values) {
+                               const gtsam::Values& values) {
```
Comment (marcusabate) [inline RobustPGO/RobustSolver.cpp]: Same as above.

```suggestion
    const gtsam::Values& values) {
```

File: RobustPGO/RobustSolver.cpp [hunk 3]
```diff
@@ -122,7 +130,8 @@ void RobustSolver::forceUpdate(const gtsam::NonlinearFactorGraph& nfg,
   optimize();
 }
 
-void RobustSolver::addOdometry(const gtsam::NonlinearFactorGraph& odom_factor, const gtsam::Values& odom_values) {
+void RobustSolver::addOdometry(const gtsam::NonlinearFactorGraph& odom_factor, 
+                               const gtsam::Values& odom_values) {
```
Comment (marcusabate) [inline RobustPGO/RobustSolver.cpp]:
```suggestion
    const gtsam::Values& odom_values) {
```

File: RobustPGO/RobustSolver.cpp [hunk 4]
```diff
@@ -133,33 +142,35 @@ void RobustSolver::addOdometry(const gtsam::NonlinearFactorGraph& odom_factor, c
   }
 }
 
-void RobustSolver::updateBatch(gtsam::NonlinearFactorGraph factors,
-    const gtsam::Values& values, const gtsam::Key& key0) {
+void RobustSolver::updateBatch(const gtsam::NonlinearFactorGraph& factors,
+                               const gtsam::Values& values,
```
Comment (marcusabate) [inline RobustPGO/RobustSolver.cpp]:
```suggestion
    const gtsam::Values& values, 
```

File: RobustPGO/RobustSolver.cpp [hunk 5]
```diff
@@ -133,33 +142,35 @@ void RobustSolver::addOdometry(const gtsam::NonlinearFactorGraph& odom_factor, c
   }
 }
 
-void RobustSolver::updateBatch(gtsam::NonlinearFactorGraph factors,
-    const gtsam::Values& values, const gtsam::Key& key0) {
+void RobustSolver::updateBatch(const gtsam::NonlinearFactorGraph& factors,
+                               const gtsam::Values& values, 
+                               const gtsam::Key& key0) {
```
Comment (marcusabate) [inline RobustPGO/RobustSolver.cpp]:
```suggestion
    const gtsam::Key& key0) {
```

File: RobustPGO/RobustSolver.cpp [hunk 6]
```diff
@@ -133,33 +142,35 @@ void RobustSolver::addOdometry(const gtsam::NonlinearFactorGraph& odom_factor, c
...
   gtsam::Key current_key = key0; // initial key
   // note that as of now only deal with between factors)
   // first load the odometry
   // order a nonlinear factor graph as odometry first
   bool extracted_odom = false;
+  gtsam::NonlinearFactorGraph update_factors = factors; 
   while (!extracted_odom) {
     bool end_of_odom = true;
-    for (size_t i = 0; i < factors.size(); i++) {
+    for (size_t i = 0; i < update_factors.size(); i++) {
```
Comment (marcusabate) [inline RobustPGO/RobustSolver.cpp]: We can make this a range-based for loop if we are using c++11:
```c++
for (gtsam::Pose3& factor : update_factors) {
```
I'm not actually sure if `gtsam::NonlinearFactorGraph` is iterable. If it is, probably the type needs to be the `gstam::Values` base and not `gtsam::Pose3`. Up to you @yunzc if this is worth doing.

File: RobustPGO/RobustSolver.h [hunk 7]
```diff
@@ -57,13 +46,14 @@ class RobustSolver : public GenericSolver{
    *  - values: linearization point for new variables
    */
   void forceUpdate(const gtsam::NonlinearFactorGraph& nfg=gtsam::NonlinearFactorGraph(),
-              const gtsam::Values& values=gtsam::Values());
+                   const gtsam::Values& values=gtsam::Values());
```
Comment (marcusabate) [inline RobustPGO/RobustSolver.h]:
```suggestion
    const gtsam::Values& values=gtsam::Values());
```

File: RobustPGO/RobustSolver.h [hunk 8]
```diff
@@ -57,13 +46,14 @@ class RobustSolver : public GenericSolver{
...
 
   /*! \brief add an odometry edge
    * odom_factor: a gtsam::NonlinearFactorGraph with a single factor
    * odom_values: a gtsam::Values with a single value
    */
-  void addOdometry(const gtsam::NonlinearFactorGraph& odom_factor, const gtsam::Values& odom_values);
+  void addOdometry(const gtsam::NonlinearFactorGraph& odom_factor, 
+                   const gtsam::Values& odom_values);
```
Comment (marcusabate) [inline RobustPGO/RobustSolver.h]:
```suggestion
    const gtsam::Values& odom_values);
```

File: RobustPGO/RobustSolver.h [hunk 9]
```diff
@@ -72,8 +62,9 @@ class RobustSolver : public GenericSolver{
    *  - values: linearization point of graph to be connected
    *  - key0: Lowest key of the graph to be connected (root of odometry)
    */
-  void updateBatch(gtsam::NonlinearFactorGraph factors,
-      const gtsam::Values& values, const gtsam::Key& key0);
+  void updateBatch(const gtsam::NonlinearFactorGraph& factors,
+                   const gtsam::Values& values,
```
Comment (marcusabate) [inline RobustPGO/RobustSolver.h]:
```suggestion
    const gtsam::Values& values, 
```

File: RobustPGO/RobustSolver.h [hunk 10]
```diff
@@ -72,8 +62,9 @@ class RobustSolver : public GenericSolver{
    *  - values: linearization point of graph to be connected
    *  - key0: Lowest key of the graph to be connected (root of odometry)
    */
-  void updateBatch(gtsam::NonlinearFactorGraph factors,
-      const gtsam::Values& values, const gtsam::Key& key0);
+  void updateBatch(const gtsam::NonlinearFactorGraph& factors,
+                   const gtsam::Values& values, 
+                   const gtsam::Key& key0);
```
Comment (marcusabate) [inline RobustPGO/RobustSolver.h]:
```suggestion
    const gtsam::Key& key0);
```

File: RobustPGO/RobustSolver.h [hunk 11]
```diff
@@ -106,8 +91,9 @@ class RobustSolver : public GenericSolver{
    *  - prior: the prior factor anchoring the graph, associated with the first value (lowest key ex. 0, or a0, etc. )
    */
   template<class T>
-  void loadGraph(const gtsam::NonlinearFactorGraph& factors, const gtsam::Values& values,
-      const gtsam::PriorFactor<T>& prior) {
+  void loadGraph(const gtsam::NonlinearFactorGraph& factors, 
+                 const gtsam::Values& values,
```
Comment (marcusabate) [inline RobustPGO/RobustSolver.h]:
```suggestion
    const gtsam::Values& values,
```

File: RobustPGO/RobustSolver.h [hunk 12]
```diff
@@ -106,8 +91,9 @@ class RobustSolver : public GenericSolver{
...
+                 const gtsam::PriorFactor<T>& prior) {
```
Comment (marcusabate) [inline RobustPGO/RobustSolver.h]:
```suggestion
    const gtsam::PriorFactor<T>& prior) {
```

File: RobustPGO/RobustSolver.h [hunk 13]
```diff
@@ -143,8 +129,9 @@ class RobustSolver : public GenericSolver{
    *    in the exisitng graph.
    */
   template<class T>
-  void addGraph(const gtsam::NonlinearFactorGraph& factors, const gtsam::Values& values,
-      const gtsam::BetweenFactor<T>& connector) {
+  void addGraph(const gtsam::NonlinearFactorGraph& factors,  
+                const gtsam::Values& values,
```
Comment (marcusabate) [inline RobustPGO/RobustSolver.h]:
```suggestion
    const gtsam::Values& values,
```

File: RobustPGO/RobustSolver.h [hunk 14]
```diff
@@ -143,8 +129,9 @@ class RobustSolver : public GenericSolver{
...
+                const gtsam::BetweenFactor<T>& connector) {
```
Comment (marcusabate) [inline RobustPGO/RobustSolver.h]:
```suggestion
    const gtsam::BetweenFactor<T>& connector) {
```

File: RobustPGO/outlier/OutlierRemoval.h [hunk 15]
```diff
@@ -32,9 +32,9 @@ class OutlierRemoval {
    *  - returns: boolean of if optimization should be called or not
    */
 	virtual bool removeOutliers(const gtsam::NonlinearFactorGraph& new_factors,
```
Comment (marcusabate) [inline RobustPGO/outlier/OutlierRemoval.h]: Weird indentation happening here. Looks like its over an indent too far.

Also move the arguments back as above.

File: RobustPGO/outlier/pcm.h [hunk 16]
```diff
@@ -208,7 +203,8 @@ class Pcm : public OutlierRemoval{
   /*! \brief goes through the loop closures and updates the corresponding adjacency matrices,
    * in preparation for max clique
    */
-  void parseAndIncrementAdjMatrix(const gtsam::NonlinearFactorGraph& new_factors, const gtsam::Values output_values){
+  void parseAndIncrementAdjMatrix(const gtsam::NonlinearFactorGraph& new_factors, 
+                                  const gtsam::Values output_values){
```
Comment (marcusabate) [inline RobustPGO/outlier/pcm.h]:
```suggestion
    const gtsam::Values output_values){
```

File: RobustPGO/outlier/pcm.h [hunk 17]
```diff
@@ -403,7 +399,7 @@ class Pcm : public OutlierRemoval{
    * inputs are 2 loop closures (a,b) and (c,d), where a,b,c,d are keys
    */
   bool areLoopsConsistent(const gtsam::BetweenFactor<poseT>& a_lcBetween_b,
-      const gtsam::BetweenFactor<poseT>& c_lcBetween_d,
+                          const gtsam::BetweenFactor<poseT>& c_lcBetween_d,
```
Comment (marcusabate) [inline RobustPGO/outlier/pcm.h]:
```suggestion
    const gtsam::BetweenFactor<poseT>& c_lcBetween_d,
```

File: RobustPGO/RobustSolver.cpp (hunk 1 above)
Comment (yunzc) [reply RobustPGO/RobustSolver.cpp]: @marcusabate  wait to double check, isn't the convention to align the argument with what is on the first line? Four spaces indentation is when there is no arguments on the first line right? https://google.github.io/styleguide/cppguide.html#Reference_Arguments

Comment (marcusabate) [reply RobustPGO/RobustSolver.cpp]: Ah you are correct, let me go through again real quick.

**Pull request merged**